### PR TITLE
Implement liftoff into grunt-cli

### PR DESCRIPTION
**Pull request by shama**
Ref: https://github.com/gruntjs/rfcs/blob/master/text/0001-liftoff.md

This is a conservative implementation of [Liftoff](https://www.npmjs.com/package/liftoff) into grunt-cli. It should be fully backwards compatible with existing Grunt versions. It needs a bit more vetting though to be sure.

This adds the ability to write your Gruntfile in any language: `Gruntfile.ts`, `Gruntfile.coffee`, `Gruntfile.babel.js` and many others. Or if the language [isnt natively supported](https://www.npmjs.com/package/interpret), you can run your own: `grunt --require custom-lang/register`.

You can also now specify v8 flags, so `grunt --harmony` gets passed to the node process (maybe not as useful these days but at least it can now).

---

Down the road, this change will allow us to remove the `coffeescript` dependency and `lib/grunt/cli.js` (among other related deps). When we do so, anyone who previously had a `Gruntfile.coffee`, would need to separately `npm install coffeescript` but the good news is they'll no longer be stuck to our old version. But all that can wait until we do a `grunt@2.0.0` release.

### CODE REVIEW

**Comment by shama** [comment]
FYI, I've been using this branch as my `grunt` command for the past 2 weeks on all my projects. Everything has been 👍. There are a few more edge cases I want to test and then will merge/release this.

File: bin/grunt [hunk 1]
```diff
@@ -4,41 +4,53 @@
...
+    aliases[short] = '--' + key;
   }
+  known[key] = gruntOptions[key].type;
+});
+
+// Parse them and return an options object.
+var options = nopt(known, aliases, process.argv, 2);
+
+if (options.version) {
+  console.log('grunt-cli v' + pkg.version);
```
Comment (Arkni) [inline bin/grunt]: I think you need to exit after this line as it will cause an exception in the line 53. Here is what I got when I run `grunt --version`:

```
$ grunt --version
grunt-cli v1.2.0

assert.js:84
  throw new assert.AssertionError({
  ^
AssertionError: missing path
    at Module.require (module.js:502:3)
    at require (internal/module.js:20:19)
    at Liftoff.<anonymous> (/home/arkni/.local/lib/node_modules/grunt-cli/bin/grunt:53:17)
    at Liftoff.execute (/home/arkni/.local/lib/node_modules/grunt-cli/node_modules/liftoff/index.js:203:12)
    at module.exports (/home/arkni/.local/lib/node_modules/grunt-cli/node_modules/flagged-respawn/index.js:51:3)
    at Liftoff.<anonymous> (/home/arkni/.local/lib/node_modules/grunt-cli/node_modules/liftoff/index.js:195:5)
    at Liftoff.<anonymous> (/home/arkni/.local/lib/node_modules/grunt-cli/node_modules/liftoff/index.js:170:7)
    at _combinedTickCallback (internal/process/next_tick.js:73:7)
    at process._tickCallback (internal/process/next_tick.js:104:9)
```

Comment (Arkni) [reply bin/grunt]: Note that I runned that command in an arbitrary folder which doesn't have `Grunt` installed to its `node_modules`.

Comment (Arkni) [reply bin/grunt]: I think that would be fixed by doing the same as the original version of the CLI: exit if the `modulePath` is not defined and the user requested the version of the CLI.

File: bin/grunt [hunk 2]
```diff
@@ -4,41 +4,53 @@
...
-var completion = require('../lib/completion');
-var info = require('../lib/info');
-var path = require('path');
-
-var basedir = process.cwd();
-var gruntpath;
-
-// Do stuff based on CLI options.
-if ('completion' in options) {
-  completion.print(options.completion);
```
Comment (Arkni) [inline bin/grunt]: Are we going to drop the `completion` command line flag? If someone was configuring the auto-completion as [pointed](https://github.com/gruntjs/grunt-cli#shell-tab-auto-completion) in the `README`, they will only see `Grunt` trying to run their tasks directly after opening a new terminal session, which will result in an exception simular to the one in [my other comment](https://github.com/gruntjs/grunt-cli/pull/117/files#r172367757).

File: bin/grunt (hunk 1 above)
Comment (shama) [reply bin/grunt]: Sweet, fixed. Thanks for the review!

File: package.json [hunk 3]
```diff
@@ -15,14 +15,15 @@
...
   "dependencies": {
-    "findup-sync": "~0.3.0",
-    "grunt-known-options": "~1.1.0",
-    "nopt": "~3.0.6",
-    "resolve": "~1.1.0"
+    "grunt-known-options": "^1.1.0",
+    "interpret": "^1.1.0",
+    "liftoff": "^2.5.0",
+    "nopt": "^4.0.1",
+    "v8flags": "^3.0.1"
```
Comment (vladikoff) [inline package.json]: are you sure you want to change from `~` to `^`?

**Comment by tkellen** [comment]
@shama, @vladikoff, @Arkni, ya'll are saints. Thank you for working on this!

File: package.json (hunk 3 above)
Comment (shama) [reply package.json]: Fixed! Thanks!

**Comment by shama** [comment]
One part I'm having second thoughts about is whether to support all the v8 flags. At the moment, [v8flags](https://www.npmjs.com/package/v8flags) needs to shell out `node --v8-options` to get the list of options. This adds a tiny perf hit in the beginning and takes a lot of options from userland.

The alternative is someone could do `node --harmony ./node_modules/.bin/grunt --my-flag` if they needed to run node with a specific flag.

**Comment by vladikoff** [comment]
> a tiny perf hit

If it's a really tiny hit that's fine.

**Comment by tkellen** [comment]
@shama That perf hit is only incurred once, we cache the output and read it from a file after subsequent runs.

**Comment by jorrit** [comment]
Is anything holding this issue back? It sounds great!

**Comment by vladikoff** [comment]
Soon.
